File: fixtures/unit-native/android/src/test/kotlin/com/hotupdater/TestHelpers.kt
package com.hotupdater

import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import okio.Buffer
import java.io.File
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import kotlin.test.assertTrue

// MARK: - Test Constants

object TestConstants {
    const val validBundleHash = "9a885c0ebee4f7a9dce994f626b1fb4cebfde6e3608fb01f714061d7c4e70e3f"
    const val corruptedBundleHash = "38893dade3c03e3521f5750c4a8ee90cd6d7b1eeb30b410a0cce483ea6ede84b"
    const val invalidBundleHash = "accc5fb6b024d45a87a6013f3aff7ddd94de4463bfd7d3814d37e090d4fd594f"
    const val appVersion = "1.0.0"
    const val fingerprint = "test-fingerprint"
    const val channel = "production"
    const val bundleId = "test-bundle-1"
}

// MARK: - Temporary Directory Management

class TempDirectoryManager {
    private val tempDirectories = mutableListOf<File>()

    fun createTempDirectory(prefix: String = "HotUpdaterTest"): File {
        val tempDir = File.createTempFile(prefix, "").apply {
            delete()
            mkdirs()
        }
        tempDirectories.add(tempDir)
        return tempDir
    }

    fun cleanupAll() {
        tempDirectories.forEach { dir ->
            dir.deleteRecursively()
        }
        tempDirectories.clear()
    }
}

// MARK: - Mock Web Server Helper

class MockWebServerHelper {
    private val server = MockWebServer()
    private var started = false

    fun start() {
        if (!started) {
            server.start()
            started = true
        }
    }

    fun shutdown() {
        if (started) {
            server.shutdown()
            started = false
        }
    }

    fun url(path: String = "/"): String = server.url(path).toString()

    fun enqueueFile(file: File, statusCode: Int = 200) {
        val buffer = Buffer()
        buffer.write(file.readBytes())
        val response = MockResponse()
            .setResponseCode(statusCode)
            .setBody(buffer)
            .addHeader("Content-Length", file.length().toString())
        server.enqueue(response)
    }

    fun enqueueNetworkError() {
        server.enqueue(MockResponse().setResponseCode(500))
    }

    fun enqueueData(data: ByteArray, statusCode: Int = 200, contentLength: Long? = null) {
        val buffer = Buffer()
        buffer.write(data)
        val response = MockResponse()
            .setResponseCode(statusCode)
            .setBody(buffer)

        contentLength?.let {
            response.addHeader("Content-Length", it.toString())
        } ?: run {
            response.addHeader("Content-Length", data.size.toString())
        }

        server.enqueue(response)
    }
}

// MARK: - Test Preferences Service

class TestPreferencesService : PreferencesService {
    private val storage = mutableMapOf<String, String>()
    private var isolationKey: String = ""

    fun configure(isolationKey: String) {
        this.isolationKey = isolationKey
    }

    private fun prefixedKey(key: String): String {
        require(isolationKey.isNotEmpty()) { "PreferencesService not configured" }
        return "$isolationKey$key"
    }

    override fun setItem(key: String, value: String?) {
        val fullKey = prefixedKey(key)
        if (value != null) {
            storage[fullKey] = value
        } else {
            storage.remove(fullKey)
        }
    }

    override fun getItem(key: String): String? {
        val fullKey = prefixedKey(key)
        return storage[fullKey]
    }
}

// MARK: - Test File System Service

class TestFileSystemService(private val documentsDir: File) : FileSystemService {
    override fun documentsPath(): String = documentsDir.absolutePath

    override fun fileExists(path: String): Boolean = File(path).exists()

    override fun createDirectory(path: String): Boolean {
        return try {
            File(path).mkdirs()
            true
        } catch (e: Exception) {
            false
        }
    }

    override fun removeItem(path: String) {
        File(path).deleteRecursively()
    }

    override fun moveItem(srcPath: String, dstPath: String) {
        val src = File(srcPath)
        val dst = File(dstPath)
        if (!src.renameTo(dst)) {
            throw Exception("Failed to move $srcPath to $dstPath")
        }
    }

    override fun copyItem(srcPath: String, dstPath: String) {
        val src = File(srcPath)
        val dst = File(dstPath)
        src.copyRecursively(dst, overwrite = true)
    }

    override fun contentsOfDirectory(path: String): List<String> {
        return File(path).list()?.toList() ?: emptyList()
    }

    override fun attributesOfItem(path: String): Map<String, Any> {
        val file = File(path)
        return mapOf(
            "size" to file.length(),
            "lastModified" to file.lastModified()
        )
    }
}

// MARK: - Progress Tracker

class ProgressTracker {
    private val progressValues = mutableListOf<Double>()

    @Synchronized
    fun track(progress: Double) {
        progressValues.add(progress)
    }

    @Synchronized
    fun reset() {
        progressValues.clear()
    }

    @Synchronized
    fun getValues(): List<Double> = progressValues.toList()

    val lastProgress: Double?
        @Synchronized get() = progressValues.lastOrNull()

    val minProgress: Double?
        @Synchronized get() = progressValues.minOrNull()

    val maxProgress: Double?
        @Synchronized get() = progressValues.maxOrNull()

    val count: Int
        @Synchronized get() = progressValues.size
}

// MARK: - File Assertions

object FileAssertions {
    fun assertFileExists(path: String, message: String = "Expected file to exist at: $path") {
        assertTrue(File(path).exists(), message)
    }

    fun assertFileNotExists(path: String, message: String = "Expected file to not exist at: $path") {
        assertTrue(!File(path).exists(), message)
    }

    fun assertDirectoryExists(path: String, message: String = "Expected directory to exist at: $path") {
        val file = File(path)
        assertTrue(file.exists() && file.isDirectory, message)
    }

    fun assertFileContains(path: String, expectedContent: String, message: String = "Expected file to contain content") {
        val content = File(path).readText()
        assertTrue(content.contains(expectedContent), message)
    }

    fun assertBundleStructure(bundlePath: String, platform: String = "android") {
        assertDirectoryExists(bundlePath)
        val indexBundle = File(bundlePath, "index.$platform.bundle")
        assertFileExists(indexBundle.absolutePath)
    }
}

// MARK: - Async Test Utilities

suspend fun <T> awaitResult(
    timeoutMs: Long = 10000,
    block: (complete: (T) -> Unit) -> Unit
): T {
    val latch = CountDownLatch(1)
    var result: T? = null
    var error: Throwable? = null

    block { value ->
        result = value
        latch.countDown()
    }

    val completed = latch.await(timeoutMs, TimeUnit.MILLISECONDS)
    if (!completed) {
        throw AssertionError("Timeout waiting for result after ${timeoutMs}ms")
    }

    error?.let { throw it }
    return result ?: throw AssertionError("Result was null")
}

// MARK: - Test Resources

object TestResources {
    fun getResourceFile(name: String): File {
        val classLoader = TestResources::class.java.classLoader
            ?: throw IllegalStateException("ClassLoader not found")

        val resource = classLoader.getResource(name)
            ?: throw IllegalArgumentException("Resource not found: $name")

        return File(resource.file)
    }

    fun getResourcePath(name: String): String {
        return getResourceFile(name).absolutePath
    }
}

File: fixtures/unit-native/ios/Tests/HotUpdaterTests/Helpers/TestHelpers.swift
import Foundation
import Testing

// MARK: - Test Constants

enum TestConstants {
    static let validBundleHash = "9a885c0ebee4f7a9dce994f626b1fb4cebfde6e3608fb01f714061d7c4e70e3f"
    static let corruptedBundleHash = "38893dade3c03e3521f5750c4a8ee90cd6d7b1eeb30b410a0cce483ea6ede84b"
    static let invalidBundleHash = "accc5fb6b024d45a87a6013f3aff7ddd94de4463bfd7d3814d37e090d4fd594f"

    static let mockBundleUrl = "https://mock.server/test-bundle.zip"
    static let appVersion = "1.0.0"
    static let fingerprint = "test-fingerprint"
    static let channel = "production"
    static let bundleId = "test-bundle-1"
}

// MARK: - Temporary Directory Management

final class TempDirectoryManager {
    private var tempDirectories: [URL] = []

    /// Create a temporary directory for testing
    func createTempDirectory(prefix: String = "HotUpdaterTest") -> URL {
        let tempDir = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(prefix)-\(UUID().uuidString)")

        try! FileManager.default.createDirectory(
            at: tempDir,
            withIntermediateDirectories: true
        )

        tempDirectories.append(tempDir)
        return tempDir
    }

    /// Clean up all temporary directories
    func cleanupAll() {
        for tempDir in tempDirectories {
            try? FileManager.default.removeItem(at: tempDir)
        }
        tempDirectories.removeAll()
    }

    deinit {
        cleanupAll()
    }
}

// MARK: - Mock Data Generators

struct MockData {
    /// Generate UpdateInfo for testing
    static func createUpdateInfo(
        bundleId: String = TestConstants.bundleId,
        appVersion: String = TestConstants.appVersion,
        fileUrl: String = TestConstants.mockBundleUrl,
        fileHash: String? = TestConstants.validBundleHash,
        fingerprint: String = TestConstants.fingerprint,
        channel: String = TestConstants.channel
    ) -> [String: Any] {
        var info: [String: Any] = [
            "bundleId": bundleId,
            "appVersion": appVersion,
            "fileUrl": fileUrl,
            "fingerprint": fingerprint,
            "channel": channel
        ]

        if let fileHash = fileHash {
            info["fileHash"] = fileHash
        }

        return info
    }

    /// Generate BundleMetadata for testing
    static func createBundleMetadata(
        bundleId: String = TestConstants.bundleId,
        version: String = "1",
        timestamp: Int64? = nil
    ) -> [String: Any] {
        return [
            "bundleId": bundleId,
            "version": version,
            "timestamp": timestamp ?? Int64(Date().timeIntervalSince1970)
        ]
    }
}

// MARK: - File Assertions

struct FileAssertions {
    /// Assert that a file exists at the given path
    static func assertFileExists(
        _ path: String,
        sourceLocation: SourceLocation = #_sourceLocation
    ) {
        #expect(
            FileManager.default.fileExists(atPath: path),
            "Expected file to exist at: \(path)",
            sourceLocation: sourceLocation
        )
    }

    /// Assert that a file does not exist at the given path
    static func assertFileNotExists(
        _ path: String,
        sourceLocation: SourceLocation = #_sourceLocation
    ) {
        #expect(
            !FileManager.default.fileExists(atPath: path),
            "Expected file to not exist at: \(path)",
            sourceLocation: sourceLocation
        )
    }

    /// Assert that a directory exists at the given path
    static func assertDirectoryExists(
        _ path: String,
        sourceLocation: SourceLocation = #_sourceLocation
    ) {
        var isDirectory: ObjCBool = false
        let exists = FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory)
        #expect(
            exists && isDirectory.boolValue,
            "Expected directory to exist at: \(path)",
            sourceLocation: sourceLocation
        )
    }

    /// Assert that a file contains expected content
    static func assertFileContains(
        _ path: String,
        expectedContent: String,
        sourceLocation: SourceLocation = #_sourceLocation
    ) throws {
        let content = try String(contentsOfFile: path, encoding: .utf8)
        #expect(
            content.contains(expectedContent),
            "Expected file at \(path) to contain '\(expectedContent)'",
            sourceLocation: sourceLocation
        )
    }

    /// Assert that a bundle directory has the correct structure
    static func assertBundleStructure(
        _ bundlePath: String,
        platform: String = "ios",
        sourceLocation: SourceLocation = #_sourceLocation
    ) {
        assertDirectoryExists(bundlePath, sourceLocation: sourceLocation)

        // Check for index bundle file
        let indexBundle = (bundlePath as NSString)
            .appendingPathComponent("index.\(platform).bundle")
        assertFileExists(indexBundle, sourceLocation: sourceLocation)
    }
}

// MARK: - Test Resource Loading

struct TestResources {
    /// Get the path to a test resource file
    static func path(for resourceName: String) -> String? {
        // In Swift Testing, resources are accessed via Bundle
        let components = resourceName.split(separator: ".")
        guard let name = components.first,
              let ext = components.last else {
            return nil
        }

        return Bundle.module.path(
            forResource: String(name),
            ofType: String(ext)
        )
    }

    /// Load test resource data
    static func data(for resourceName: String) throws -> Data {
        guard let resourcePath = path(for: resourceName) else {
            throw NSError(
                domain: "TestResources",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "Resource not found: \(resourceName)"]
            )
        }
        return try Data(contentsOf: URL(fileURLWithPath: resourcePath))
    }

    /// Get URL for test resource
    static func url(for resourceName: String) -> URL? {
        guard let resourcePath = path(for: resourceName) else {
            return nil
        }
        return URL(fileURLWithPath: resourcePath)
    }
}

// MARK: - Progress Tracking

final class ProgressTracker {
    private(set) var progressValues: [Double] = []
    private let lock = NSLock()

    func track(_ progress: Double) {
        lock.lock()
        defer { lock.unlock() }
        progressValues.append(progress)
    }

    func reset() {
        lock.lock()
        defer { lock.unlock() }
        progressValues.removeAll()
    }

    var lastProgress: Double? {
        lock.lock()
        defer { lock.unlock() }
        return progressValues.last
    }

    var minProgress: Double? {
        lock.lock()
        defer { lock.unlock() }
        return progressValues.min()
    }

    var maxProgress: Double? {
        lock.lock()
        defer { lock.unlock() }
        return progressValues.max()
    }
}

// MARK: - Async Test Utilities

struct AsyncTestUtils {
    /// Wait for a condition to be true with timeout
    static func wait(
        timeout: TimeInterval = 5.0,
        condition: @escaping () -> Bool
    ) async throws {
        let startTime = Date()
        while !condition() {
            if Date().timeIntervalSince(startTime) > timeout {
                throw NSError(
                    domain: "AsyncTestUtils",
                    code: 408,
                    userInfo: [NSLocalizedDescriptionKey: "Timeout waiting for condition"]
                )
            }
            try await Task.sleep(nanoseconds: 100_000_000) // 100ms
        }
    }
}
